### PR TITLE
Strategies Pattern Improvement 

### PR DESCRIPTION
**Pull request by SamueLagunju**

### CODE REVIEW

File: crypto/__main__.py [hunk 1]
```diff
@@ -68,19 +68,20 @@ def main(encrypt, decrypt, files):
 
     # NEW - File encryption based on its extension
     elif files:
-        for filename in files:
+        for file in files:
+            filename, file_extension = file.split('.', 1)
             crypter = Crypter(filename, SeanStrategy())
```
Comment (KaiPrince) [inline crypto/__main__.py:73]: The Crypter class never uses this filename. In fact, right now it's basically an empty class. I think it should either be given more responsibilities, or removed.

File: crypto/__main__.py [hunk 2]
```diff
@@ -68,19 +68,20 @@ def main(encrypt, decrypt, files):
...
-            if any(file in filename for file in encryption_extension):
-                print("Encrypting: {0}".format(filename))
+            if file_extension in encryption_extension:
+                print("Encrypting: {0}".format(file))
                 output_text = crypter.encrypt_txt(file_contents)
-            write_file(filename, output_text)
-            check_write(output_text, filename)
-            new_file = convert_ext(filename)
-            print("Output File: {0}".format(new_file))
+            new_file_name = crypter.convert_ext(filename, file_extension)
```
Comment (KaiPrince) [inline crypto/__main__.py:81]: Be careful, there are two places in the other if clauses where you forgot to repoint this function.

I think it's indicative of a larger issue: you have too much code duplication. Consider that this elif clause is just a combination of the two above it, but there is no code reuse between them.

Two possible remedies are to extract the common logic to new functions, or to push the logic down into the Crypter class.

File: crypto/crypter.py [hunk 3]
```diff
@@ -40,3 +40,6 @@ def encrypt_txt(self, plain_text):
     # RETURNS       :   decrypted text
     def decrypt_txt(self, cipher_text):
         return self.strategy.decrypt_text(cipher_text)
+
+    def convert_ext(self, file_name, old_ext):
```
Comment (KaiPrince) [inline crypto/crypter.py:44]: If you end up moving logic into this class, you may find this method unnecessary. Or, at least, worth making private.

File: crypto/__main__.py [hunk 4]
```diff
@@ -68,19 +68,20 @@ def main(encrypt, decrypt, files):
...
+        for file in files:
+            filename, file_extension = file.split('.', 1)
             crypter = Crypter(filename, SeanStrategy())
-            file_contents = read_file(filename)
-            if any(file in filename for file in decryption_extension):
-                print("Decrypting: {0}".format(filename))
+            file_contents = read_file(file)
+            if file_extension in decryption_extension:
+                print("Decrypting: {0}".format(file))
                 output_text = crypter.decrypt_txt(file_contents)
```
Comment (KaiPrince) [inline crypto/__main__.py:77]: You may find that you need to rename the decrypt_txt family of functions in both Crypter and Strategy to reflect the new possibility of non-ascii (binary) file data.

**Comment by KaiPrince** [comment]
In summary, it seems like we've opened a can of worms. It may be time to re-think and ask "what is the Crypter class? Why does it exist, and what's its responsibility? Is it a do-er or a coordinator?". This is totally normal when refactoring; it often requires you to critically re-examine the structures you had previously decided on. Good work on extending the strategy class.

**Review by KaiPrince — CHANGES_REQUESTED** [review]
In summary, it seems like we've opened a can of worms. It may be time to re-think and ask "what is the Crypter class? Why does it exist, and what's its responsibility? Is it a do-er or a coordinator?". This is totally normal when refactoring; it often requires you to critically re-examine the structures you had previously decided on. Good work on extending the strategy class.

**Review by SamueLagunju — COMMENTED** [review]
Will be applying main logic to crypter class

File: crypto/__main__.py [hunk 5]
```diff
@@ -21,70 +19,24 @@ def main(encrypt, decrypt, files):
...
+    rules = [encrypt,
+             decrypt,
+             files]
+    # If there are files
+    if any(rules):
+        for element in rules:
+            if element:
+                for file in element:
+                    if validate_file(file):
+                        list_of_files.append(file)
```
Comment (KaiPrince) [inline crypto/__main__.py:32]:
```suggestion
    for element in rules:
        for file in element:
            if validate_file(file):
                list_of_files.append(file)
```

c'mon man...

Comment (SamueLagunju) [reply crypto/__main__.py:32]: late night programming 🤦🏾‍♂️

**Review by SamueLagunju — COMMENTED** [review]
GOAT Pull request merge into mediocre PR

**Review by SamueLagunju — COMMENTED** [review]
Changes have been implemented. I think we've refactored enough of the project

**Review by SamueLagunju — COMMENTED** [review]
Tidying Up.

**Pull request merged**